utils/tps/src/main.rs

```diff
@@ -18,39 +37,184 @@ struct Args {
+			for evt in events.iter() {
+				let evt = evt?;
+				let event_name = evt.variant_name();
+				if event_name == "CandidateIncluded" {
+					// If the CandidateIncluded event occurs, we need to get the CandidateDescriptor by decoding bytes
+					let mut values = evt.field_bytes();
+					let candidate_receipt = CandidateReceipt::<Hash>::decode(&mut values)?;
+					let descriptor = candidate_receipt.descriptor();
+					info!(
+						"ParaBlock Subscriber ===> New ParaHead: {:?} for ParaId: {:?}",
+						descriptor.para_head, descriptor.para_id
+					);
+					tx.send(descriptor.para_head).await?;
+				}
+			}
+		}
+	}
+	Ok(())
 }
 
-pub async fn calc_tps(api: &Api, n: usize) -> Result<(), Error> {
+/// in case we're monitoring TPS on a parachain
+/// calc_para_tps thread listens for which parachain block to scrape
+async fn calc_para_tps(
+	para_api: &Api,
+	mut rx: Receiver<H256>,
+	default_parablock_time: u64,
+) -> Result<(), Error> {
 	let storage_timestamp_storage_addr = runtime::storage().timestamp().now();
+	let mut trx_in_parablock = 0;
+	while let Some(para_head) = rx.recv().await {
+		info!("TPS Counter ===> Received ParaHead: {:?}", para_head);
+		let parablock = para_api.blocks().at(Some(para_head)).await?;
+		let parabody = parablock.body().await?;
+		let parablock_number = parablock.number();
+		let previous_parablock_number = parablock_number - 1;
+		let maybe_previous_parablock_hash =
+			para_api.rpc().block_hash(Some(previous_parablock_number.into())).await?;
 
-	let block_1_hash = api.rpc().block_hash(Some(1u32.into())).await?;
+		// Need to handle the case where we cannot get the previous parablock timestamp
+		let parablock_time = match maybe_previous_parablock_hash {
+			Some(hash) => {
+				let parablock_timestamp = para_api
+					.storage()
+					.fetch(&storage_timestamp_storage_addr, Some(para_head))
+					.await?
+					.unwrap();
+				let previous_parablock_timestamp = para_api
+					.storage()
+					.fetch(&storage_timestamp_storage_addr, Some(hash))
+					.await?
+					.unwrap();
+				let time_diff = parablock_timestamp - previous_parablock_timestamp;
+				info!("TPS Counter ===> Parablock time estimated at: {:?}", time_diff);
+				time_diff
+			},
+			// Assume default if unable to get the previous parablock from parablock number
+			None => {
+				warn!(
+					"TPS Counter ===> Assuming default parablock time of: {:?}",
+					default_parablock_time
+				);
+				Duration::as_secs_f64(&Duration::new(default_parablock_time, 0)) as u64
+			},
+		};
+		for extrinsic in parabody.extrinsics() {
+			for events in extrinsic.events().await {
+				for event in events.iter() {
+					let evt = event?;
+					let variant = evt.variant_name();
+					if variant == "Transfer" {
+						trx_in_parablock += 1;
+					}
+				}
+			}
+		}
+		// Print to stdout and reset counter
+		info!(
+			"TPS Counter ===> Counted {} TPS in ParaHead: {:?}",
+			trx_in_parablock / parablock_time,
+			para_head
+		);
+	}
+	trx_in_parablock = 0;
```

Comment:
shouldn't we be resetting `trx_in_parablock` inside the `while` loop?

Reply:
Yes I believe we should, indeed.